lib/Dialect/FIRRTL/FIRRTLTypes.cpp

```diff
@@ -1042,14 +1060,46 @@ std::pair<size_t, bool> FVectorType::rootChildFieldID(size_t fieldID,
                         fieldID >= childRoot && fieldID <= rangeEnd);
 }
 
+//===----------------------------------------------------------------------===//
+// RefType
+//===----------------------------------------------------------------------===//
+
+namespace circt {
+namespace firrtl {
+namespace detail {
+struct RefTypeStorage : mlir::TypeStorage {
+  using KeyTy = FIRRTLBaseType;
+
+  RefTypeStorage(KeyTy value) : value(value) {}
+
+  bool operator==(const KeyTy &key) const { return key == value; }
+
+  static RefTypeStorage *construct(TypeStorageAllocator &allocator, KeyTy key) {
+    return new (allocator.allocate<RefTypeStorage>()) RefTypeStorage(key);
+  }
+
+  KeyTy value;
+};
+
+} // namespace detail
+} // namespace firrtl
+} // namespace circt
+
+auto RefType::get(FIRRTLBaseType type) -> RefType {
+  assert(type.isGround() && "reference base type must be ground");
```

Comment:
Is this ever reachable given the verifiers on reference types?

Reply:
Looks like constructing a type calls its verifier anyway, so we'll get the "assert valid" behavior regardless, at least if we define such a hook.
Not sure if reachable, that's why there's an assert for ensuring only valid RefType's are constructed.  But will drop it since I think `verify` does all I was trying to do more directly anyway.